fields.go

```diff
@@ -145,7 +145,7 @@ func makeStructFields(root reflect.Type) (structFields, *SemanticError) {
 				switch {
 				case tf == jsontextValueType:
 					f.fncs = nil // specially handled in arshal_inlined.go
-				case tf.Kind() == reflect.Map && tf.Key() == stringType:
+				case tf.Kind() == reflect.Map && tf.Key().Kind() == reflect.String:
```

Comment:
What happens if `tf.Key()` implements `encoding.TextMarshaler`, `MarshalerV1` or `MarshalerV2`?

Reply:
At mininum, we should forbit any named `string` type that can marshal itself, but that does open up the door to problematic situations where someone adds a marshaler to named string type after the fact, causing this to no longer be marshalable. However, I think that situation is okay as it's generally not safe to add a marshaler after the fact.

Reply:
> What happens if tf.Key() implements encoding.TextMarshaler, MarshalerV1 or MarshalerV2?

Any custom marshalers/unmarshalers were ignored for inlined map keys.

I agree that it makes more sense to forbid key types that implement JSON marshal or unmarshal methods. I've added that restriction and included a couple of tests.

> that does open up the door to problematic situations where someone adds a marshaler to named string type after the fact, causing this to no longer be marshalable. However, I think that situation is okay as it's generally not safe to add a marshaler after the fact.

FWIW, this is currently the case for inlined structs as well. So, the proposed change remains consistent with the existing implementation.